### PR TITLE
Replace Guava cache with Caffeine cache for vertex cache

### PR DESCRIPTION
**Pull request by ministat**
Fix the issue #3185

Guava cache has performance bug in 100% read scenario. The solution is to use Caffeine cache which is also suggested by Guava community. See google/guava#2408 for more details. Meanwhile, Caffeine cache beats Guava cache in all scenarios.

The following result is for read scenario.

Benchmark (cacheType) Mode Cnt Score Error Units
VertexCacheBenchmark.run guava avgt 5 0.001 ± 0.001 ms/op
VertexCacheBenchmark.run caffeine avgt 5 ≈ 10⁻⁴ ms/op

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

**Comment by porunov** [comment]
Thank you for the contribution @ministat , @clovertrail !
Replacing Guava cache with Caffeine cache definitely makes sense to me.
That said, to accept your contribution you need to sign CLA to give JanusGraph permission to use the contribution you created.
Would you be able to sign EasyCLA?
Signing a CLA might be a problem if you signed CLA for one GitHub account but made a commit using information from another GitHub account.
In your case I see that the commit is made by @clovertrail but the PR is opened by @ministat .
In this case you need to use `clovertrail` as a GitHub account when signing EasyCLA.
If you intended to use `ministat` account then you will need to change the author of the commit to use name and email of `ministat` GitHub account.
Below I will show possible solutions on how to correctly sign the commit in your case.
1) Solution 1 - the case when you want to use `ministat` account only. In this case we need to update commit with the correct author information and sign the commit (by using `-s` parameter).
```
git config user.name "Hongjiang Zhang"
git config user.email "your.email@example.org"
git commit --amend -s --author="Hongjiang Zhang <your.email@example.org>"
git push -f
```
After that login into GitHub with `ministat` username. And click the next link to sign CLA:
https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/1543822/77385607/3188/#/?version=2

2) Solution 2 - the case when you want to use `clovertrail` account only.
```
git config user.name "Your Full Name"
git config user.email "email.of.clovertrail@example.org"
git commit --amend -s
git push -f
```
After that login into GitHub with `clovertrail` username. And click the next link to sign CLA:
https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/1543822/77385607/3188/#/?version=2

Please, let me know if you have any problems with signing EasyCLA, and I will help you with signing it.

**Comment by ministat** [comment]
@porunov Thanks for your suggestion. I have passed the CLA checked. Appreciated!

**Review by porunov — COMMENTED** [review]
Checking the dependency tree I see that we get Caffeine 2.3.1 from TinkerPop dependency (`org.apache.tinkerpop:gremlin-groovy:3.5.4`).
I'm just thinking if we should use the latest Caffeine 3.1.1 or should we stick to the one provided by TinkerPop. Do you know if there were any breaking changes? If no, then I would prefer trying out the latest Caffeine dependency 3.1.1.

**Comment by ben-manes** [comment]
@porunov The main reason for the major version bump was that v3.x requires Java 11 or higher, whereas previously it was Java 8. This allowed the implementation code to switch from sun.misc.Unsafe to VarHandles (fully removed in 3.0.2). Otherwise the changes were minor clean ups that had little impact on users (see [release notes](https://github.com/ben-manes/caffeine/releases/tag/v3.0.0)).

**Comment by porunov** [comment]
> @porunov The main reason for the major version bump was that v3.x requires Java 11 or higher, whereas previously it was Java 8. This allowed the implementation code to switch from sun.misc.Unsafe to VarHandles (fully removed in 3.0.2). Otherwise the changes were minor clean ups that had little impact on users (see [release notes](https://github.com/ben-manes/caffeine/releases/tag/v3.0.0)).

Thank you for the clarification! JanusGraph supports Java 11 but we still write everything in Java 8 to support clients on Java 8, so I think in this case it should be safe to use Caffeine 2.9.3.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java
Comment (porunov) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java:68]: I see that cache statistics is always enabled here, even so many applications won't use it. It's definitely helpful for analyses but many times it won't be needed. Moreover, I don't know if there is any memory / performance overhead with having statistics enabled. 
What I noticed is that you report statistics only when closing cache and only when debug is enabled. Thus, I would recommend here to enable `.recordStats()` only when debug log level is enabled. I.e. something like this:
```
Caffeine<Long, InternalVertex> cacheBuilder = Caffeine.newBuilder().maximumSize(maxCacheSize)
    .removalListener(new CaffeineRemovalListener())
    .executor(Runnable::run); // according to the https://github.com/ben-manes/caffeine/discussions/757
if(log.isDebugEnabled()){
    cacheBuilder = cacheBuilder.recordStats();
}
cache = cacheBuilder.build();
```
Another option (maybe even better but not necessary to do this in this PR) is to add a new configuration option to JanusGraph to enable / disable cache stats recording.

**Review by porunov — COMMENTED** [review]
The PR looks good but I have small nitpicks below.

Could you please add explicit Caffeine dependency to root `pom.xml` like the next:
```
<dependency>
    <groupId>com.github.ben-manes.caffeine</groupId>
    <artifactId>caffeine</artifactId>
    <version>${caffeine.version}</version>
</dependency>
```
Also, specify `<caffeine.version>2.9.3</caffeine.version>`.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java:68]: Also, in case you do this with debug enabled / disabled then make sure in `close` method there is no NPE thrown by any chance (I just didn't check if that is even possible). But for additional protection you can wrap that code in additional if.
I.e.:
```
if(log.isDebugEnabled()){
    long end = System.currentTimeMillis();
    CacheStats stats = cache.stats();
    log.debug("Caffeine cache (lifespan: {}ms) stats: {}", end - createdTime, stats);  
}
```

Comment (ben-manes) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java:68]: I haven't benchmarked this for a while, but have recently been advising a developer writing a similar cache for .NET. He recently ported Java's `LongAdder` to replace an equivalent of `AtomicLong` for statistics, since I mentioned that was a likely bottleneck. You can see how there is a degradation but the penalty is not noticeable in an application as the performance is very high. (https://github.com/bitfaster/BitFaster.Caching/pull/219)

**Comment by porunov** [comment]
@ministat ignore failed Hbase and dist tests. Looks like there is a problem with getting docker image for some reason but it's unrelated to your changes

**Comment by ministat** [comment]
Add explicit Caffeine dependency, and excludes the lower version of org.checkerframework:checker-qual from guava, otherwise, there are many dependency convergence errors like the following:
```
Dependency convergence error for org.checkerframework:checker-qual:jar:3.12.0:compile paths to dependency are:
+-org.janusgraph:janusgraph-dist:pom:1.0.0-SNAPSHOT
  +-org.apache.tinkerpop:hadoop-gremlin:jar:3.5.4:compile
    +-com.google.guava:guava:jar:31.1-jre:compile
      +-org.checkerframework:checker-qual:jar:3.12.0:compile
and
+-org.janusgraph:janusgraph-dist:pom:1.0.0-SNAPSHOT
  +-org.apache.tinkerpop:gremlin-groovy:jar:3.5.4:compile
    +-com.github.ben-manes.caffeine:caffeine:jar:2.9.3:compile
      +-org.checkerframework:checker-qual:jar:3.19.0:compile

[WARNING] Rule 0: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for org.checkerframework:checker-qual:3.12.0 paths to dependency are:
+-org.janusgraph:janusgraph-dist:1.0.0-SNAPSHOT
  +-org.apache.tinkerpop:hadoop-gremlin:3.5.4
    +-com.google.guava:guava:31.1-jre (managed) <-- com.google.guava:guava:16.0.1
      +-org.checkerframework:checker-qual:3.12.0
and
+-org.janusgraph:janusgraph-dist:1.0.0-SNAPSHOT
  +-org.apache.tinkerpop:gremlin-groovy:3.5.4
    +-com.github.ben-manes.caffeine:caffeine:2.9.3 (managed) <-- com.github.ben-manes.caffeine:caffeine:2.3.1
      +-org.checkerframework:checker-qual:3.19.0
```

**Comment by ministat** [comment]
@porunov Thanks for you comments, I have updated my patch.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java
Comment (porunov) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java:68]: Looks like you forgot to remove `.recordStats()` here. I.e. you are enabling it 2 times, here and below.

Comment (ministat) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java]: My fault!

Comment (ministat) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java]: My fault!

Comment (ministat) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java]: Fixed

**Comment by porunov** [comment]
@ministat I fixed HBase tests issue in the latest commit in `master` and `v0.6` branches. Would you be able to rebase your branch against the latest `master` commit please to ensure that all tests pass?
I.e.
```
git remote add upstream https://github.com/JanusGraph/janusgraph
git fetch upstream master
git rebase upstream/master
git push -f
```

**Comment by porunov** [comment]
@ministat it also looks like the benchmark class is often stuck in the first or second iteration for long time. I'm not sure but it could be flaky. I was waiting for 10 minutes for the second iteration to pass but it didn't finish. Sometimes it hangs on the warmup iteration. Do you know what is a correct way to run the benchmark you included?

**Review by porunov — APPROVED** [review]
LGTM. Thank you @ministat !
The PR looks good to me but I still suggest to rebase it against the latest `master` commit just in case (described [here](https://github.com/JanusGraph/janusgraph/pull/3188#issuecomment-1231447860)).

**Comment by ministat** [comment]
@porunov, I have rebased per your suggestion.
For the benchmark, I run it through idea firstly. In order to run it in cmd line, I copied the idea's java classpath when it was running, pasted the long classpath into a file and run it again. I got the same perf results. I did not see hang in my test.

**Comment by porunov** [comment]
> @porunov, I have rebased per your suggestion. For the benchmark, I run it through idea firstly. In order to run it in cmd line, I copied the idea's java classpath when it was running, pasted the long classpath into a file and run it again. I got the same perf results. I did not see hang in my test.

Thanks! Now to merge this PR we need to either wait for another commiter to approve this PR or wait 7 days to merge this PR with lazy consensus.

**Comment by ministat** [comment]
@porunov I found the build "mvn clean package" failed in any janusgraph-xxx folder, for example, janusgraph-benchmark.
Before my CI, the build is successful. I'm still investigating. Do you have any suggestion?
```
~/Work/OpenSource/ministat_janusgraph/janusgraph-core @ LM-SHC-16507917 (hongjizhang) 
| => mvn clean package
[INFO] Scanning for projects...
[INFO] 
[INFO] -------------------< org.janusgraph:janusgraph-core >-------------------
[INFO] Building JanusGraph-Core: Core Library for JanusGraph 1.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:3.2.0:clean (default-clean) @ janusgraph-core ---
[INFO] Deleting /Users/hongjizhang/Work/OpenSource/ministat_janusgraph/janusgraph-core/target
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0:enforce (enforce-dependency-convergence) @ janusgraph-core ---
[WARNING] 
Dependency convergence error for org.checkerframework:checker-qual:jar:3.12.0:compile paths to dependency are:
+-org.janusgraph:janusgraph-core:jar:1.0.0-SNAPSHOT
  +-org.janusgraph:janusgraph-driver:jar:1.0.0-SNAPSHOT:compile
    +-com.google.guava:guava:jar:31.1-jre:compile
      +-org.checkerframework:checker-qual:jar:3.12.0:compile
and
+-org.janusgraph:janusgraph-core:jar:1.0.0-SNAPSHOT
  +-com.github.ben-manes.caffeine:caffeine:jar:2.9.3:compile
    +-org.checkerframework:checker-qual:jar:3.19.0:compile

[WARNING] Rule 1: org.apache.maven.plugins.enforcer.DependencyConvergence failed with message:
Failed while enforcing releasability. See above detailed error message.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.278 s
[INFO] Finished at: 2022-08-30T23:13:27+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0:enforce (enforce-dependency-convergence) on project janusgraph-core: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Comment by ministat** [comment]
After I "mvn install -DskipTests" from the root, the "mvn clean package" from janusgraph-core is successful. So, I think the previous failure is caused by referencing the wrong library.

**Comment by li-boxuan** [comment]
>  I would suggest excluding this dependency from guava which is in janusgraph-driver and adding checker-qual:3.19.0 explicitly to the pom.xml

I think we should add `checker-qual:3.19.0` to the `<dependencyManagement>` section in root `pom.xml`. We don't want to introduce this dependency explicitly since we are not using it.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java:322]: Can you also remove `GuavaVertexCache` class since it's not being used anymore?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java:42]: (nitpick): I would remove this comment since your code perfectly explains itself.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java:131]: (nitpick) Redundant `toString()` call.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java:44]: (nitpick) It is a common practice in JanusGraph to put private classes at the bottom of the file.

**Review by li-boxuan — COMMENTED** [review]
Thanks for your contribution! I just have some nitpicks:

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java:322]: As far as I understand it is used in the benchmark test class. I guess we could move that class into testing module then

Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java:322]: I guess we can remove that class with the benchmark test class in the follow up PRs but I'm OK either way.

Comment (ben-manes) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java:322]: Should the `GuavaVertexCache` be moved there rather than have it linger around as dead code?

Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java:322]: Actually I think it's better to leave the benchmark test class in this PR for historical reasons. We can move `GuavaVertexCache` to test module and then later after the full transition to Caffeine cache is done we can remove those test classes with the benchmark classes. It's just good if those benchmark classes are in git history

Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java:322]: >  later after the full transition to Caffeine cache is done we can remove those test classes with the benchmark classes

That sounds good to me!

Comment (ministat) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java:322]: I agree to leave GuavaVertexCache there. I'm ok to compose several following PRs to clean it up.
Meanwhile, I found there are some other Guava cache, like https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/GuavaSubqueryCache.java, and it is supposed to be replaced by Caffeine cache also, but I have not yet benchmark it.

Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java:322]: > I agree to leave GuavaVertexCache there. I'm ok to compose several following PRs to clean it up. Meanwhile, I found there are some other Guava cache, like https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/GuavaSubqueryCache.java, and it is supposed to be replaced by Caffeine cache also, but I have not yet benchmark it.

There are 8 caches in JanusGraph actually and all of them need to be replaced by Caffeine (not in the current PR but in follow up PRs). I posted those places here: https://github.com/JanusGraph/janusgraph/issues/871#issuecomment-1230207577

File: janusgraph-core/pom.xml
Comment (li-boxuan) [inline janusgraph-core/pom.xml:87]: (nitpick) You don't need this line because you have already specified the version in the `<dependencyManagement>` section of root `pom.xml`.

**Review by li-boxuan — APPROVED** [review]
Neat! I am okay with the failing codecov.

File: janusgraph-core/pom.xml
Comment (porunov) [reply janusgraph-core/pom.xml:87]: Right. I.e. this is enough when the version has been already specified in `dependencyManagement`:
```
<dependency>
            <groupId>com.github.ben-manes.caffeine</groupId>
            <artifactId>caffeine</artifactId>
</dependency>
```

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java
Comment (mad) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java:39]: Caffeine has special feature for avoid this extra field - https://github.com/ben-manes/caffeine/issues/264

Implementation could be looks like https://gist.github.com/mad/9119f48812000a03dea16c7485c621fc

@ministat @porunov

Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java:39]: > Caffeine has special feature for avoid this extra field - [ben-manes/caffeine#264](https://github.com/ben-manes/caffeine/issues/264)
> 
> Implementation could be looks like https://gist.github.com/mad/9119f48812000a03dea16c7485c621fc
> 
> @ministat @porunov

That's interesting! I would prefer refactoring this class to eliminate this extra field as you proposed. That said, as it is a performance related part of the code we should probably execute benhmarking for any improvement we make here, just to make sure we don't have a regression. Other than that your proposal is much cleaner, so that's great.
Should we open a separate ticket to track this?

Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java:39]: Also, a related transitioning PR which replaces other Guava caches with Caffeine cache is: #3198 
After that transitioning is merged I believe we need to refactor [CaffeineVertexCache.java](https://github.com/JanusGraph/janusgraph/blob/c30c7071227de338a63d82336b34f749f2384b22/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java) and [ExpirationKCVSCache.java](https://github.com/JanusGraph/janusgraph/blob/0822cce3d7a50cdb530d13a363f076b18f9b14be/janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/cache/ExpirationKCVSCache.java).

**Pull request merged**
